packages/eslint-plugin/src/rules/no-misused-promises.ts

```diff
@@ -486,13 +486,40 @@ function isFunctionParam(
   return false;
 }
 
-// Get the positions of parameters which are void functions (and not also
+function checkThenableOrVoid(
+  checker: ts.TypeChecker,
+  node: ts.CallExpression | ts.NewExpression,
+  type: ts.Type,
+  index: number,
+  thenableReturnIndices: Set<number>,
+  voidReturnIndices: Set<number>,
+): void {
+  if (isThenableReturningFunctionType(checker, node.expression, type)) {
+    thenableReturnIndices.add(index);
+  } else if (isVoidReturningFunctionType(checker, node.expression, type)) {
+    // If a certain argument accepts both thenable and void returns,
+    // a promise-returning function is valid
+    if (!thenableReturnIndices.has(index)) {
+      voidReturnIndices.add(index);
+    }
+  }
+}
+
+// Get the positions of arguments which are void functions (and not also
 // thenable functions). These are the candidates for the void-return check at
 // the current call site.
-function voidFunctionParams(
+// If the function parameters end with a 'rest' parameter, then we consider
+// the array type parameter (e.g. '...args:Array<SomeType>') when determining
+// if trailing arguments are candidates.
+function voidFunctionArguments(
   checker: ts.TypeChecker,
   node: ts.CallExpression | ts.NewExpression,
 ): Set<number> {
+  // 'new' can be used without any arguments, as in 'let b = new Object;'
+  // In this case, there are no argument positions to check, so return early.
+  if (!node.arguments) {
+    return new Set<number>();
+  }
```

Comment:
Great code addition, but missing test coverage - could you add one in please?

Reply:
I meant for this to be tested by `new TakeCallbacks;` and `new TakesVoidCb;` but it looks like running `yarn lint --fix` changed it. I've added `// eslint-disable-next-line @typescript-eslint/internal/plugin-test-formatting` to allow it to pass the internal lint.